### PR TITLE
[fpmsyncd]: Add support for SRv6

### PR DESCRIPTION
**Pull request by cscarpitta**
HLD: https://github.com/sonic-net/SONiC/pull/1620

**What I did**

Extended fpmsyncd to process Netlink TLVs containing SRv6 information (SRv6 local SIDs and routes).

**Why I did it**

fpmsyncd did not process Netlink TLVs containing SRv6 information (SRv6 local SIDs and routes).

**How I verified it**

This PR implements new set of test cases for SRv6 `test_srv6.py`
The sonic-mgmt tests for SRv6 are added in this PR: https://github.com/sonic-net/sonic-mgmt/pull/15510

### CODE REVIEW

**Comment by eddieruan-alibaba** [comment]
@cscarpitta can you rebase it to fix the compile issue? https://github.com/sonic-net/sonic-buildimage/pull/18715 is merged.

**Comment by eddieruan-alibaba** [comment]
This change has been cherry-picked into phoenix wing and used in daily run.

http://phoenixwing.com.cn/build

**Review by eddieruan-alibaba — APPROVED** [review]

**Comment by AlanYoush** [comment]
@cscarpitta 
Hi cscarpitta
I find that there is another PR: https://github.com/sonic-net/sonic-swss/pull/2515 and it is also uesd to support SRv6. What is the difference between them both? 
And I want to use srv6 feature in SONiC-202311, so I should merge SRv6 PR code into SONiC-202311, and which PR Code should I merge ?
Thanks.

**Review by ahsalam — APPROVED** [review]
This changes are needed for programming SRv6 from FRR.

File: fpmsyncd/routesync.cpp [hunk 1]
```diff
@@ -584,13 +873,576 @@ void RouteSync::onEvpnRouteMsg(struct nlmsghdr *h, int len)
...
+        {
+            parseEncapSrv6SteerRoute(tb[RTA_ENCAP], vpn_sid, src_addr);
+        }
+        SWSS_LOG_DEBUG("Rx MsgType:%d encap:%d vpn_sid:%s src_addr:%s",
+                        h->nlmsg_type, encap, vpn_sid.c_str(),
+                        src_addr.c_str());
+
+        if (vpn_sid.empty())
+        {
+            return false;
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply fpmsyncd/routesync.cpp:899]: Added an error message:
`SWSS_LOG_ERROR("Received an invalid SRv6 route: vpn_sid is empty");`

Thanks.

File: fpmsyncd/routesync.cpp [hunk 2]
```diff
@@ -81,6 +145,8 @@ RouteSync::RouteSync(RedisPipeline *pipeline) :
     m_vnet_routeTable(pipeline, APP_VNET_RT_TABLE_NAME, true),
     m_vnet_tunnelTable(pipeline, APP_VNET_RT_TUNNEL_TABLE_NAME, true),
     m_warmStartHelper(pipeline, &m_routeTable, APP_ROUTE_TABLE_NAME, "bgp", "bgp"),
+    m_srv6LocalSidTable(pipeline, APP_SRV6_MY_SID_TABLE_NAME, true),
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply fpmsyncd/routesync.cpp]: Thanks for the comment. I changed localSid -> mySid everywhere in the code.

File: tests/mock_tests/fpmsyncd/receive_srv6_localsids_ut.cpp [hunk 3]
```diff
@@ -0,0 +1,1203 @@
...
+
+        struct nlmsg *nl_obj = create_srv6_localsid_nlmsg(RTM_NEWSRV6LOCALSID, &_localsid, _block_len, _node_len, _func_len, _arg_len, _action, NULL, &_adj);
+        if (!nl_obj)
+            throw std::runtime_error("SRv6 Local SID creation failed");
+
+        /* Send the Netlink object to the FpmLink */
+        m_fpmLink->processRawMsg(&nl_obj->n);
+
+        /* Check that fpmsyncd created the correct entries in APP_DB */
+        std::string action;
```
Comment (kperumalbfn) [inline tests/mock_tests/fpmsyncd/receive_srv6_localsids_ut.cpp:123]: Could you pls create a wrapper to create MYSID table key by passing block/node/func/arg lengths and local SID? That will be very useful as this is used in many places.

File: tests/test_srv6.py [hunk 4]
```diff
@@ -627,6 +629,369 @@ def test_srv6(self, dvs, testlog):
...
+        self.setup_srv6(dvs)
+
+        # configure srv6 usid locator
+        dvs.runcmd("vtysh -c \"configure terminal\" -c \"segment-routing\" -c \"srv6\" -c \"locators\" -c \"locator loc1\" -c \"prefix fc00:0:2::/48 block-len 32 node-len 16 func-bits 16\" -c \"behavior usid\"")
+
+        # create srv6 mysid udt6 behavior
+        dvs.runcmd("ip -6 route add fc00:0:2:ff05::/128 encap seg6local action End.DT6 vrftable {} dev sr0".format(self.vrf_table_id))
+
+        # check application database
+        self.pdb.wait_for_entry("SRV6_MY_SID_TABLE", "32:16:16:0:fc00:0:2:ff05::")
```
Comment (kperumalbfn) [inline tests/test_srv6.py:964]: Could you check Endpoint behavior of MYSID entry in APPL_DB?

File: tests/test_srv6.py [hunk 5]
```diff
@@ -627,6 +629,369 @@ def test_srv6(self, dvs, testlog):
...
+    def test_AddRemoveSrv6MySidUDT6(self, dvs, testlog):
+
+        _, output = dvs.runcmd(f"vtysh -c 'show zebra dplane providers'")
+        if 'dplane_fpm_sonic' not in output:
+            pytest.skip("'dplane_fpm_sonic' required for this test is not available, skipping", allow_module_level=True)
+
+        self.setup_srv6(dvs)
+
+        # configure srv6 usid locator
+        dvs.runcmd("vtysh -c \"configure terminal\" -c \"segment-routing\" -c \"srv6\" -c \"locators\" -c \"locator loc1\" -c \"prefix fc00:0:2::/48 block-len 32 node-len 16 func-bits 16\" -c \"behavior usid\"")
```
Comment (kperumalbfn) [inline tests/test_srv6.py:958]: In FRR, is this the only extra config('behavior usid') to determine SID is uSID?

File: tests/test_srv6.py [hunk 6]
```diff
@@ -627,6 +629,369 @@ def test_srv6(self, dvs, testlog):
         assert nexthop_entries == get_exist_entries(self.adb.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP")
         assert route_entries == get_exist_entries(self.adb.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
 
+
+class TestSrv6MySidFpmsyncd(object):
```
Comment (kperumalbfn) [inline tests/test_srv6.py:633]: @cscarpitta Thanks for adding extensive UTs for all the endpoint behaviors. Could you also add a SRV6 steering routes and check for SRV6 nexthop type - SAI_NEXT_HOP_TYPE_SRV6_SIDLIST in ASIC_DB?

Comment (kperumalbfn) [reply tests/test_srv6.py:633]: Please use FPMSyncd to add the route and check the DB entries.

**Comment by kperumalbfn** [comment]
@dgsudharsan Please check

File: tests/mock_tests/fpmsyncd/receive_srv6_localsids_ut.cpp (hunk 3 above)
Comment (cscarpitta) [reply tests/mock_tests/fpmsyncd/receive_srv6_localsids_ut.cpp]: Done. I added a function `get_srv6_my_sid_table_key()` that builds the MYSID table key and change the code to use this function.

File: tests/test_srv6.py (hunk 4 above)
Comment (cscarpitta) [reply tests/test_srv6.py:970]: Done. I added some checks to verify endpoint behavior + attributes of the MYSID entries in APPL_DB

File: tests/test_srv6.py (hunk 6 above)
Comment (cscarpitta) [reply tests/test_srv6.py:633]: Done.

I added two test cases to verify SRv6 steering routes:
* test_AddRemoveSrv6SteeringRouteIpv4
* test_AddRemoveSrv6SteeringRouteIpv6

**Comment by cscarpitta** [comment]
@kperumalbfn 

Many thanks for your review. All comments have been addressed.

File: tests/test_srv6.py (hunk 5 above)
Comment (cscarpitta) [reply tests/test_srv6.py:964]: Yes, in FRR there is a 'behavior usid' command that sets a flag "uSID" in the locator data structure. When "uSID" flag is set, SIDs allocated are uSIDs

**Comment by eddieruan-alibaba** [comment]
@cscarpitta can you rebase it to latest master branch? You have "This branch is out-of-date with the base branch". We need to pass all sanity checks.

**Comment by kperumalbfn** [comment]
@cscarpitta please check the test failures

**Comment by kperumalbfn** [comment]
@cscarpitta Please resolve all the SRV6 related failures

image

**Comment by kperumalbfn** [comment]
@eddyk-nvidia @dgsudharsan Fpmsyncd changes to program my_sid entries.

**Comment by kperumalbfn** [comment]
@cscarpitta Please include all the related PRs in the description.

**Comment by kperumalbfn** [comment]
@cscarpitta Any scale number for SRV6 steer route and MYSID add/delete with this fpmsyncd. Please share some numbers.

File: tests/test_srv6.py [hunk 7]
```diff
@@ -627,6 +630,923 @@ def test_srv6(self, dvs, testlog):
...
+        self.setup_srv6(dvs)
+
+        # enable VRF strict mode
+        dvs.runcmd("sysctl -w net.vrf.strict_mode=1")
+
+        # configure srv6 usid locator
+        dvs.runcmd("vtysh -c \"configure terminal\" -c \"segment-routing\" -c \"srv6\" -c \"locators\" -c \"locator loc1\" -c \"prefix fc00:0:2::/48 block-len 32 node-len 16 func-bits 16\" -c \"behavior usid\"")
+
+        # create srv6 mysid udt4 behavior
+        dvs.runcmd("ip -6 route add fc00:0:2:ff05::/128 encap seg6local action End.DT4 vrftable {} dev sr0".format(self.vrf_table_id))
```
Comment (kperumalbfn) [inline tests/test_srv6.py:1169]: @cscarpitta To support TUNNEL DSCP/TTL mode MY_SID object has tunnel_id. Currently, there is no orchagent support for this. But how do we add this in FRR/Fpmsyncd?

Comment (cscarpitta) [reply tests/test_srv6.py:1169]: Hi @kperumalbfn To support the DSCP/TTL, we can extend FRR to provide these parameters to fpmsyncd and extend orchagent to process them

File: fpmsyncd/routesync.cpp [hunk 8]
```diff
@@ -45,6 +47,65 @@ using namespace swss;
...
+#define DEFAULT_SRV6_MY_SID_NODE_LEN "16"
+#define DEFAULT_SRV6_MY_SID_FUNC_LEN "16"
+#define DEFAULT_SRV6_MY_SID_ARG_LEN "0"
+
+enum srv6_localsid_action {
+	SRV6_LOCALSID_ACTION_UNSPEC				= 0,
+	SRV6_LOCALSID_ACTION_END				= 1,
+	SRV6_LOCALSID_ACTION_END_X				= 2,
+	SRV6_LOCALSID_ACTION_END_T				= 3,
+	SRV6_LOCALSID_ACTION_END_DX2			= 4,
```
Comment (kperumalbfn) [inline fpmsyncd/routesync.cpp:60]: Current SAI doesn't support L2 XConnect, does FRR support this?

File: fpmsyncd/routesync.cpp [hunk 9]
```diff
@@ -144,6 +207,226 @@ void RouteSync::parseEncap(struct rtattr *tb, uint32_t &encap_value, string &rma
...
+    uint32_t action_buf = SRV6_LOCALSID_ACTION_UNSPEC;
+    char vrf_buf[IFNAMSIZ + 1] = {0};
+    char adj_buf[MAX_ADDR_SIZE + 1] = {0};
+
+    if (tb[SRV6_LOCALSID_FORMAT])
+    {
+        if (!parseSrv6MySidFormat(tb[SRV6_LOCALSID_FORMAT], block_len,
+                                node_len, func_len, arg_len))
+        {
+            SWSS_LOG_ERROR("Invalid Srv6 MySid format");
```
Comment (kperumalbfn) [inline fpmsyncd/routesync.cpp:386]: could you add details about node/block/function in the error log message

File: fpmsyncd/routesync.cpp [hunk 10]
```diff
@@ -144,6 +207,226 @@ void RouteSync::parseEncap(struct rtattr *tb, uint32_t &encap_value, string &rma
...
+        inet_ntop(AF_INET, nh4, adj_buf, MAX_ADDR_SIZE);
+    }
+
+    if (tb[SRV6_LOCALSID_VRFNAME])
+    {
+        memcpy(vrf_buf, (char *)RTA_DATA(tb[SRV6_LOCALSID_VRFNAME]),
+               strlen((char *)RTA_DATA(tb[SRV6_LOCALSID_VRFNAME])));
+    }
+
+    action = mySidAction2Str(action_buf);
```
Comment (kperumalbfn) [inline fpmsyncd/routesync.cpp:418]: unknown action, return false?

File: fpmsyncd/routesync.cpp (hunk 9 above)
Comment (cscarpitta) [reply fpmsyncd/routesync.cpp]: Added. Thanks for the comment.

File: fpmsyncd/routesync.cpp (hunk 10 above)
Comment (cscarpitta) [reply fpmsyncd/routesync.cpp:421]: Done. Thanks for the comment.

File: fpmsyncd/routesync.cpp (hunk 8 above)
Comment (cscarpitta) [reply fpmsyncd/routesync.cpp:60]: No, currently FRR does not support L2 XConnect.

**Comment by eddieruan-alibaba** [comment]
@cscarpitta can you rebase your branch ? "This branch is out-of-date with the base branch"

**Comment by cscarpitta** [comment]
> @cscarpitta can you rebase your branch ? "This branch is out-of-date with the base branch"

Done.

**Review by kperumalbfn — COMMENTED** [review]
@cscarpitta please include sonic-mgmt test PR(https://github.com/sonic-net/sonic-mgmt/pull/15510) in the description.

**Comment by kperumalbfn** [comment]
Changes LGTM

**Review by kperumalbfn — APPROVED** [review]

**Comment by cscarpitta** [comment]
> @cscarpitta please include sonic-mgmt test PR([sonic-net/sonic-mgmt#15510](https://github.com/sonic-net/sonic-mgmt/pull/15510)) in the description.

@kperumalbfn Thanks for the review. I updated the PR description.

**Pull request merged**